src/execution/index/art/art.cpp

```diff
@@ -204,6 +204,163 @@ void ART::GenerateKeys(DataChunk &input, vector<unique_ptr<Key>> &keys) {
+		vector<KeySection> child_sections;
+		GetChildSections(child_sections, keys, key_section);
+
+		auto node_type = Node::GetTypeBySize(child_sections.size());
+		Node::New(node_type, node);
+
+		auto prefix_length = key_section.depth - prefix_start;
+		node->prefix = Prefix(start_key, prefix_start, prefix_length);
+
+		// recurse on each child section
+		for (auto &child_section : child_sections) {
+			Node *new_child = nullptr;
+			Construct(keys, row_ids, new_child, child_section, has_constraint);
+			Node::InsertChild(node, child_section.key_byte, new_child);
+		}
+	}
+}
+
+void FindFirstNotNullKey(vector<unique_ptr<Key>> &keys, bool &skipped_all_nulls, idx_t &start_idx) {
+
+	if (!skipped_all_nulls) {
+		for (idx_t i = 0; i < keys.size(); i++) {
+			if (keys[i]) {
+				start_idx = i;
+				skipped_all_nulls = true;
+				return;
+			}
+		}
+	}
+}
+
+void ART::ConstructAndMerge(IndexLock &lock, PayloadScanner &scanner, Allocator &allocator) {
+
+	auto payload_types = logical_types;
+	payload_types.emplace_back(LogicalType::ROW_TYPE);
+
+	auto skipped_all_nulls = false;
+	auto temp_art = make_unique<ART>(this->column_ids, this->unbound_expressions, this->constraint_type, this->db);
+	for (;;) {
+		DataChunk ordered_chunk;
+		ordered_chunk.Initialize(allocator, payload_types);
+		ordered_chunk.SetCardinality(0);
+		scanner.Scan(ordered_chunk);
+		if (ordered_chunk.size() == 0) {
+			break;
+		}
+
+		// get the key chunk and the row_identifiers vector
+		DataChunk row_id_chunk;
+		ordered_chunk.Split(row_id_chunk, ordered_chunk.ColumnCount() - 1);
+		auto &row_identifiers = row_id_chunk.data[0];
+
+		D_ASSERT(row_identifiers.GetType().InternalType() == ROW_TYPE);
+		D_ASSERT(logical_types[0] == ordered_chunk.data[0].GetType());
+
+		// generate the keys for the given input
+		vector<unique_ptr<Key>> keys;
+		GenerateKeys(ordered_chunk, keys);
+
+		// we order NULLS FIRST, so we might have to skip nulls at the start of our sorted data
+		idx_t start_idx = 0;
+		FindFirstNotNullKey(keys, skipped_all_nulls, start_idx);
+
+		if (start_idx != 0 && IsPrimary()) {
+			throw ConstraintException("NULLs in new data violate the primary key constraint of the index");
+		}
+
+		if (!skipped_all_nulls) {
+			if (IsPrimary()) {
+				// chunk consists only of NULLs
+				throw ConstraintException("NULLs in new data violate the primary key constraint of the index");
+			}
+			continue;
+		}
+
+		// prepare the row_identifiers
+		row_identifiers.Flatten(ordered_chunk.size());
+		auto row_ids = FlatVector::GetData<row_t>(row_identifiers);
+
+		// construct the ART of this chunk
+		auto art = make_unique<ART>(this->column_ids, this->unbound_expressions, this->constraint_type, this->db);
+		auto key_section = KeySection(start_idx, ordered_chunk.size() - 1, 0, 0);
+		auto has_constraint = IsPrimary() || IsUnique();
+		Construct(keys, row_ids, art->tree, key_section, has_constraint);
```

Comment:
Then I need to put `KeySection` in the header file, no? Which should be fine. But I am not sure if the logic here makes sense, because `Construct` is called recursively on nodes, not on ART.

So maybe a private function of a node? But then `KeySection` needs to be public? Maybe in `art_key.hpp`?
What do you think?

Reply:
Yeah, fair argument, let it be.